lib/service/awsoidc.go

```diff
@@ -0,0 +1,263 @@
+	if !process.Config.Proxy.Enabled {
+		return nil
+	}
+
+	// start process only after teleport process has started
+	if _, err := process.WaitForEvent(process.GracefulExitContext(), TeleportReadyEvent); err != nil {
+		return nil
+	}
+	process.log.Infof("The new service has started successfully. Checking for deploy service updates every %v.", updateDeployAgentsInterval)
+
+	resp, err := process.getInstanceClient().Ping(process.GracefulExitContext())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !resp.ServerFeatures.AutomaticUpgrades {
+		return nil
+	}
+
+	periodic := interval.New(interval.Config{
+		Duration: updateDeployAgentsInterval,
+		Jitter:   retryutils.NewSeventhJitter(),
+	})
+	defer periodic.Stop()
+
+	for {
+		if err := process.updateDeployServiceAgents(process.GracefulExitContext(), process.getInstanceClient()); err != nil {
+			process.log.Warningf("Update failed: %v. Retrying in ~%v", err, updateDeployAgentsInterval)
+		}
+
+		select {
+		case <-periodic.Next():
+		case <-process.GracefulExitContext().Done():
+			return nil
+		}
+	}
+}
+
+func (process *TeleportProcess) updateDeployServiceAgents(ctx context.Context, authClient *auth.Client) error {
+	cmc, err := authClient.GetClusterMaintenanceConfig(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var criticalEndpoint string
+	if automaticupgrades.GetChannel() != "" {
+		criticalEndpoint, err = url.JoinPath(automaticupgrades.GetChannel(), "critical")
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	critical, err := automaticupgrades.Critical(process.GracefulExitContext(), criticalEndpoint)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !withinUpgradeWindow(cmc, process.Clock) && !critical {
+		return nil
+	}
+
+	teleportVersion, err := getStableTeleportVersion(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	issuer, err := awsoidc.IssuerFromPublicAddress(process.proxyPublicAddr().Addr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	clusterNameConfig, err := authClient.GetClusterName()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	clusterName := clusterNameConfig.GetClusterName()
+
+	databases, err := authClient.GetDatabases(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	awsRegions := make(map[string]interface{})
+	for _, database := range databases {
+		if database.IsAWSHosted() && database.IsRDS() {
+			awsRegions[database.GetAWS().Region] = nil
+		}
+	}
+
+	var resources []types.Integration
+	var nextKey string
+	for {
+		igs, nextKey, err := authClient.ListIntegrations(ctx, 0, nextKey)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		resources = append(resources, igs...)
+		if nextKey == "" {
+			break
+		}
+	}
+
+	limit := rate.NewLimiter(rate.Every(updateDeployAgentsRateLimit), 1)
+	for _, ig := range resources {
+		spec := ig.GetAWSOIDCIntegrationSpec()
+		if spec == nil {
+			continue
+		}
+		integrationName := ig.GetName()
+
+		for region := range awsRegions {
```

Comment:
It looks like we're iterating over every integration times every region. Do we have information about which database was discovered by which integration so we can only iterate over regions that actually have databases for a particular integration?

Reply:
Not that I know of. @marcoandredinis?

Reply:
No, we don't include anything that could backtrack the integration used.
I don't expect people to have a lot of integrations, but I still think we should improve here.

What do you think about adding a label?
It might bring false positives, because they might have a self-hosted DatabaseService proxying the connection and not using the ECS Agent.
Still an improvement 👍 
Let me know what you think

Reply:
> What do you think about adding a label?

Would this mean adding a label to the database?
Is this something that should be included in this PR?

Reply:
No, I don't think it is worth it.
Let's keep the current scope